service/dps/storm/topologies/ic/ic-topology/src/main/java/eu/europeana/cloud/service/dps/storm/topologies/ic/topology/api/ImageConverterServiceImpl.java

```diff
@@ -53,58 +59,42 @@
-                String fileName = urlParams.get(ParamConstants.P_FILENAME);
-                String cleanName = FilenameUtils.removeExtension(fileName);
-                String inputExtension = ExtensionHelper.getExtension(TaskTupleUtility.getParameterFromTuple(stormTaskTuple, PluginParameterKeys.MIME_TYPE));
-                long randomValue = UUID.randomUUID().getMostSignificantBits();
-                String tempFileName = String.valueOf(randomValue);
-                folderPath = persistStreamToTemporaryStorage(inputStream, tempFileName, inputExtension);
-                String inputFilePath = buildFilePath(folderPath, tempFileName, inputExtension);
+                String inputMimeType = TaskTupleUtility.getParameterFromTuple(stormTaskTuple, PluginParameterKeys.MIME_TYPE);
+                assert (TIFF_MIME_TYPE.equals(inputMimeType) || JPEG_MIME_TYPE.equals(inputMimeType));
+                String inputExtension = ExtensionHelper.getExtension(inputMimeType);
+                String tempFileName = UUID.randomUUID().toString();
+                folderPath = ImageConverterUtil.persistStreamToTemporaryStorage(inputStream, tempFileName, inputExtension);
+                String inputFilePath = ImageConverterUtil.buildFilePath(folderPath, tempFileName, inputExtension);
                 String outputExtension = ExtensionHelper.getExtension(TaskTupleUtility.getParameterFromTuple(stormTaskTuple, PluginParameterKeys.OUTPUT_MIME_TYPE));
-                String outputFilePath = buildFilePath(folderPath, tempFileName, outputExtension);
-                if (outputFilePath != null) {
-                    List<String> properties = new ArrayList<>();
-                    properties.add(TaskTupleUtility.getParameterFromTuple(stormTaskTuple, PluginParameterKeys.KAKADU_ARGUEMENTS));
-                    converterContext.convert(inputFilePath, outputFilePath, properties);
-                    File outputFile = new File(outputFilePath);
-                    InputStream outputStream = new FileInputStream(outputFile);
-                    stormTaskTuple.setFileData(outputStream);
-                    stormTaskTuple.addParameter(PluginParameterKeys.OUTPUT_FILE_NAME, cleanName + outputExtension);
-                    LOGGER.info("The converting process for file " + stormTaskTuple.getFileUrl() + " completed successfully");
+                String outputFilePath = ImageConverterUtil.buildFilePath(folderPath, tempFileName, outputExtension);
+
+
+                if (JPEG_EXTENSION.equals(inputExtension)) {
+                    String tempOutputFilePath = ImageConverterUtil.buildFilePath(folderPath, tempFileName, TIFF_EXTENSION);
+                    converterContext.convert(inputFilePath, tempOutputFilePath, null);
+                    inputFilePath = tempOutputFilePath;
                 }
+
+                converterContext.setConverter(new KakaduConverterTiffToJP2());

```

Comment:
It could be clear when is its used separate contexts or even re-factor whole follow to chain of responsibility pattern.